Task 967

cmcantalupo opened the pull request:
- Fixes #967 from github issues
- Adds trace generation and plotting for test_ee_short_region_slop test

cmcantalupo committed:
Add trace data to ee_short_region_slop test

- Add INSTRUCTION_RETIRED signal to trace and report
  in the run that generates trace data.
- Use geopm.hash.crc32_str() in integration
  tests for frequency map policy.
- Adding test that generates trace plots.
- Use .loc and .diff features of pandas DF
- Seperate selector for region name nd region count
- Tune the trace images to cover the same data
- Enable removal of very short regions from report plot
- Separate min max freq from queries

Change-Id: I61d287f9cf648c3954377d214dc1125d7fb70e5f
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg commented on test_integration/test_ee_short_region_slop.cpp:
```diff
@@ -50,7 +50,7 @@ int main(int argc, char **argv)
     bool is_verbose = false;
     geopm::Profile &prof = geopm::Profile::default_profile();
     // Run twelve trials with region duration ranging from 100 us - 400 ms
```
comment is no longer correct

dianarg commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -70,6 +72,13 @@
     from test_integration import util
     geopmpy.error.exc_clear()
 
+# Globals controlling uniform y axis limits in all plots
```
Could be moved into test class?  Plotting functions also would need to move

cmcantalupo replied:
I'm going to leave these separate for now with the idea that some other test may want to reuse the plotting routines.  We can make these bounds a default if that happens.

dianarg commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -204,71 +241,89 @@ def tearDownClass(cls):
+                                               (trace['REGION_COUNT'] == region_count[0])].iloc[0]
+                end_time = trace['TIME'].loc[(trace['REGION_HASH'] == end_region_hash) &
+                                             (trace['REGION_COUNT'] == region_count[-1])].iloc[-1]
+                duration = end_time - begin_time
```
duration is a confusing name
just used for width of plot

dianarg commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
+                     (trace['TIME'] < end_time)]
+
+
+def generate_trace_region_plot(trace, out_path, region_name, region_count):
```
add "overlay" to name?

dianarg commented on test_integration/test_scaling_region.py:
```diff
@@ -93,14 +94,8 @@ def setUpClass(cls):
-                            0xf3aa44cb, 0xba9639ec, 0x61d2be85, 0x28eec3a2,
-                            0xd2b7c6a6, 0x9b8bbb81, 0x40cf3ce8, 0x09f341cf,
-                            0xb1914011, 0xf8ad3d36, 0x07949283]
+        cls._region_hash = [geopmpy.hash.crc32_str('scaling_region_{}'.format(idx))
```
yay

bgeltz commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
+    di = trace['INSTRUCTIONS_RETIRED'].diff()
+    dc = trace['CYCLES_THREAD'].diff()
+    return di / dc
+
```
need another \n

bgeltz commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
+
+def get_freq(trace):
+    return trace['FREQUENCY']
+
```
need another \n

bgeltz commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
+
+def select_region(trace, region_prefix, extension=''):
+    """e.g.:
+    >>> select_region(trace, 'scaling_2')
+    >>> select_region(trace, 'scaling_', range(12))
+    >>> select_region(trace, '', ['scaling_3', 'timed_3'])
```
Nice.

bgeltz commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
+    return trace.loc[trace['REGION_HASH'].isin(region_hash)]
+
+
+def select_time_range(trace, duration, begin_time=None):
```
I like these helper funcs.  I'm wondering how we should organize a #import with all of them.

cmcantalupo replied:
Seems like a good idea for another PR.

bgeltz commented on test_integration/test_ee_short_region_slop.py:
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
+    plt.subplot(2,1,2)
+    plt.legend(freq_legend)
+    plt.savefig(out_path)
+    plt.close()
```
What do you have against \n's ? 😆

cmcantalupo replied:
I have removed the patch that has the memory alignment change.  I tried to refactor the blob into a loop, but it wasn't worth the hassle.

bgeltz reviewed (changes requested):
This is pretty slick with all the plotting stuff.  Tiny whitespace changes needed, and could you add to the scaling model region patch commit msg some more about why you're doing that?

cmcantalupo committed:
Enable control of frequency range with environment

Change-Id: I896c3897c90b463798dbcd6e7c0fc55adc8615da
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add an sbatch script for mcfly to execute the slop test

Change-Id: I5f82f60ca2d2ab469eff44852f671ee23f6d2070
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Trace temperature in test_ee_short_region_slop

- Add temperature to time series plot
- Explicitly set OMP_NUM_THREADS to 42
- Remove uncore frequency from report signals: not a
  monotone signal

Change-Id: I3d8288f8b935bb8126ec723922ddefa14f77f536
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Don't delete result data.

- localize image_path to test that uses it

Change-Id: Id2e6efa44ba77fbac14b80626e4faacfd8f87cf8
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Refactor test launch.

- Break SetUp into smaller functions.
- Add doc strings.

Change-Id: I0ef6db897db2cab87e05e173351c19e3e0e93b6b
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Double number of ranks in test_ee_short_region_slop

- Halve scaling model region foot print
- Reduce OMP thread count in sbatch script

Change-Id: I94e93b14f6ff6675990cd803b5d7b10c62d907de
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add barriers to test_ee_short_region_slop test

Change-Id: Ib5b60bf495db98af99023bc28710cd57045440b8
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix issues identified in review

Change-Id: Iad2bb8d7f54e1e7742038dd4634a357c6c1f8e50
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix issue that occurs when raw 64 bit signals are added to trace

- Raw signals end in the '#' character, but this was being used as
  a comment character by the parser.

Change-Id: If7a530ab7529f4e9be4a17794a805a32fcbf73b1
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg reviewed (dismissed):
I don't see any reason not to merge this as is now.  In the future, we might want to convert the sbatch script to a more generic launch script similar to the tutorials.

cmcantalupo committed:
Change the ScalingModelRegion array sizes to vary with ranks per node

- When run with two ranks per node on a two socket system the
  performance increases with this change.

Change-Id: I6eace8a97cf54380a4a2de7301610195e9e0b22e
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
